### PR TITLE
Navigation tweaks for Android/iOS

### PR DESCRIPTION
**Pull request by PureWeen**
#### Android
- The navigation graph now recreates from a resumed activity
- AppBarButton has a backbutton and it works
- Wired up the app activity so if there's an existing window it just uses that
- Fixed our ToNative so when trying to create a view with a new mauicontext it'll null out the previous one
- Added a NET6 solution that only targets Android for build performance reasons

#### iOS
- Fixed back button so it only navigates once

### CODE REVIEW

File: src/Core/src/Handlers/View/ViewHandlerOfT.cs [hunk 1]
```diff
@@ -128,5 +129,11 @@ internal ViewHandler()
 			get => (TVirtualView?)base.VirtualView;
 			private protected set => base.VirtualView = value;
 		}
+
+		protected TVirtualView VirtualViewWithValidation([CallerMemberName] string callerName = "")
```
Comment (PureWeen) [inline src/Core/src/Handlers/View/ViewHandlerOfT.cs:133]: This way we don't have to keep checking for null everywhere that we know the VirtualView should have a value

File: src/Core/src/Handlers/View/ViewHandlerOfT.Android.cs [hunk 2]
```diff
@@ -11,6 +13,12 @@ public partial class ViewHandler<TVirtualView, TNativeView> : INativeViewHandler
 		View? INativeViewHandler.NativeView => (View?)base.NativeView;
 		public Context? Context => MauiContext?.Context;
 
+		protected Context ContextWithValidation([CallerMemberName] string callerName = "")
```
Comment (PureWeen) [inline src/Core/src/Handlers/View/ViewHandlerOfT.Android.cs:16]: This way we don't have to keep checking for null everywhere we know that it shouldn't be

File: src/Controls/src/Core/Handlers/NavigationPage/NavigationPageHandler.Android.cs [hunk 3]
```diff
@@ -1,63 +1,69 @@
...
-using FragmentManager = AndroidX.Fragment.App.FragmentManager;
-using FragmentTransaction = AndroidX.Fragment.App.FragmentTransaction;
-using Object = Java.Lang.Object;
+using System;
 
 namespace Microsoft.Maui.Controls.Handlers
 {
 	public partial class NavigationPageHandler :
-		ViewHandler<NavigationPage, AView>//, IManageFragments, IOnClickListener, ILifeCycleState
+		ViewHandler<NavigationPage, AView>
```
Comment (PureWeen) [inline src/Controls/src/Core/Handlers/NavigationPage/NavigationPageHandler.Android.cs:19]: Most of these changes are just to make "nullable enable" work

File: src/Controls/src/Core/Handlers/NavigationPage/NavigationPageHandler.Android.cs [hunk 4]
```diff
@@ -98,9 +102,10 @@ protected override void ConnectHandler(AView nativeView)
 			}
 
 			graph.StartDestination = destinations[0];
+
 			NavHost.NavController.SetGraph(graph, null);
 
-			for (var i = 1; i < destinations.Count; i++)
+			for (var i = NativeNavigationStackCount; i < NavigationStackCount; i++)
```
Comment (PureWeen) [inline src/Controls/src/Core/Handlers/NavigationPage/NavigationPageHandler.Android.cs:108]: When you set the graph the backstack automatically recreates so we're only checking here if we need to push more pages

File: src/Controls/src/Core/Handlers/NavigationPage/NavigationPageHandler.Android.cs [hunk 5]
```diff
@@ -125,254 +130,24 @@ void OnPushed(object sender, NavigationRequestedEventArgs e)
 
 		internal void OnPop()
 		{
-			NavHost.NavController.NavigateUp();
+			VirtualViewWithValidation()
```
Comment (PureWeen) [inline src/Controls/src/Core/Handlers/NavigationPage/NavigationPageHandler.Android.cs:133]: Fire everything through the Navigation page

Life's easier when you always follow the same circle

File: src/Controls/src/Core/NavigationPage.cs [hunk 6]
```diff
@@ -315,7 +326,7 @@ async Task<Page> INavigationPageController.RemoveAsyncInner(Page page, bool anim
 			var removed = true;
 
 			EventHandler<NavigationRequestedEventArgs> requestPop = PopRequested;
-			if (requestPop != null)
+			if (requestPop != null && !requestedFromHandler)
```
Comment (PureWeen) [inline src/Controls/src/Core/NavigationPage.cs:329]: If the handler just wants you to remove a page don't fire requestPop

This eliminates all those super annoying _isPopping type checks in the renderer

File: src/Controls/src/Core/Platform/Android/MauiFragmentNavDestination.cs [hunk 7]
```diff
@@ -13,18 +13,27 @@ class MauiFragmentNavDestination : FragmentNavigator.Destination
 		public IMauiContext MauiContext => NavigationPageHandler.MauiContext;
 		public NavigationPageHandler NavigationPageHandler { get; }
 
+		// Todo we want to generate the same ids for each page so if the app is recreated
+		// we want these to match up
+		static Dictionary<IPage, int> Pages = new Dictionary<IPage, int>();
```
Comment (PureWeen) [inline src/Controls/src/Core/Platform/Android/MauiFragmentNavDestination.cs:18]: I promise I'll make this better

File: src/Core/src/Platform/Android/MauiContext.cs [hunk 8]
```diff
@@ -6,23 +6,34 @@ namespace Microsoft.Maui
 {
 	public class MauiContext : IMauiContext
 	{
-		readonly Context _context;
+		readonly WeakReference<Context> _context;
```
Comment (PureWeen) [inline src/Core/src/Platform/Android/MauiContext.cs:9]: Never store context outside of a weakreference

**Review by Redth — APPROVED** [review]

**Pull request merged**